awx/api/views/__init__.py

```diff
@@ -1445,6 +1445,37 @@ class CredentialInputSourceSubList(SubListCreateAttachDetachAPIView):
     parent_key = 'target_credential'
 
 
+class CredentialTypeExternalTest(SubDetailAPIView):
+    view_name = _('External Credential Type Test')
+    model = models.CredentialType
+    serializer_class = serializers.EmptySerializer
+
+    def post(self, request, *args, **kwargs):
+        obj = self.get_object()
+        credential_id = request.data.get('credential', None)
+
+        test_inputs = {}
+        if credential_id is not None:
+            cred = models.Credential.accessible_objects(request.user, 'use_role').filter(id=credential_id).first()
+            if cred is None or cred.credential_type.kind != 'external':
```

Comment:
So this still requires you to save the Hashi credential?  Couldn't this endpoint just take the raw `inputs` dictionary from the form _and_ the credential type ID and call `CredentialType.objects.get(pk=N).plugin(**inputs)` without having to save an actual credential?

Reply:
A saved credential isn't required. We're handling two scenarios here:

The first scenario is for testing from a completely new and unsaved credential form. In this situation, the UI doesn't send a `credential_id` in the post request and this entire block of code is skipped.

The second scenario is for testing from the edit form of an external credential that already exists. For this, we need to include the credential id in the post request. We do this because an update request to an _existing_ credential won't necessarily contain all of the information needed to run and test the plugin: For example, unless the user specifically changes an existing secret field on the form, that value for that field is sent to the API as a placeholder `$encrypted$`. Unless we replace these placeholders with their actual values from the DB, the plugin will fail to run if it needs those values.

Reply:
One thing we can do here is handle the testing of updates (scenario 2) with its own endpoint, nested under `credentials/n/test`. I went back and forth on this so I have the code laying around for both. Open to suggestions.

Reply:
Okay, I see now.  I like the idea of splitting this into separate endpoints.  Makes sense!